shell/packages/sandstorm-accounts-merge/accounts-merge-server.js

```diff
@@ -0,0 +1,241 @@
+                                            sourceIdentities: sourceIdentities,
+                                            sourceServices: oldServices}},
+                         $unset: servicesToRemove,
+                         $pull: {identities: {id: {$in: Object.keys(oldIdentities)}},
+                                 mergedUsers: destUserId} });
+  }
+});
+
+SandstormAccountsMerge.registerObservers = function (db, backend) {
+
+  Meteor.users.find({"merging.status": "pending"}).observe({
+    added: function(losingUser) {
+      console.log("merging user " + JSON.stringify(sourceUser));
+      var winningUserId = losingUser.merging.destinationUserId;
+      var winningUser = Meteor.users.findOne(winningUserId);
+
+      if (winningUser.mergedUsers.indexOf(losingUser._id) == -1) {
+        // Now let's construct a modifier that will update the new user.
+        var fieldsToSet = {};
+        var identitiesToPush = [];
+        losingUser.merging.sourceIdentities.forEach(function(identity) {
+          identitiesToPush.push(identity);
+          if (identity.service === "dev") {
+            fieldsToSet.devName = losingUser.devName;
+          } else if (identity.service === "demo") {
+            fieldsToSet.expires = losingUser.expires;
+          } else if (identity.service === "github") {
+            fieldsToSet["services.github"] = losingUser.merging.sourceServices.github;
+          } else if (identity.service === "google") {
+            fieldsToSet["services.google"] = losingUser.merging.sourceServices.google;
+          } else if (identity.service === "emailToken") {
+            fieldsToSet["services.emailToken"] = losingUser.merging.sourceServices.emailToken;
+          }
+        });
+
+        var modifier = {$push: {identities: {$each: identitiesToPush},
+                                mergedUsers: losingUser._id}};
+        if (Object.keys(fieldsToSet).length > 0) {
+          modifier["$set"] = fieldsToSet;
+        }
+        Meteor.users.update({_id: winningUserId}, modifier);
+      }
+
+      db.collections.notifications.update({userId: losingUser._id},
+                                          {$set: {userId: winningUserId}},
+                                          {multi: true});
+
+      db.collections.userActions.find({userId: losingUser._id}).forEach(function (action) {
+        var newAction = _.omit(action, "_id");
+        newAction.userId = winningUserId;
+        db.collections.userActions.upsert({userId: winningUserId, packageId: action.packageId,
+                                           title: action.title, nounPhrase: action.nounPhrase,
+                                           command: action.command},
+                                          {$set: newAction});
+      });
+
+      var grains = Grains.find({userId: losingUser._id}).fetch();
+
+      db.collections.grains.update({userId: losingUser._id},
```

Comment:
I like how you made account merging resume in case of server failure. I think you'll need to do the same at the grain level, though. Currently if there is a crash immediately after this update runs, then the grains will end up all broken, and we wont retry moving them when starting back up because their user IDs have already changed.